py/profiling.c

```diff
@@ -0,0 +1,1046 @@
+    
+    /* long path */ if (1) {
+        mp_printf(&mp_plat_print,
+            "@0x%p:%q:%q+0x%04x:%d",
+            ip,
+            prelude->qstr_source_file,
+            prelude->qstr_block_name,
+            offset,
+            prof_bytecode_lineno(rc, offset)
+        );
+    }
+    
+    /* bytecode */ if (0) {
+        mp_printf(&mp_plat_print,
+            " %02x %02x %02x %02x",
+            ip[0], ip[1], ip[2], ip[3], ip[4]
+        );
+    }
+    
+    mp_printf(&mp_plat_print,
+        " 0x%02x %q [%d]",
+        *(byte*)(ip),
+        instruction->qstr_opname,
+        instruction->arg
+    );
+    if (instruction->argobj != mp_const_none) {
+        mp_printf(&mp_plat_print, " $");
+        mp_obj_print_helper(&mp_plat_print, instruction->argobj, PRINT_REPR);
+    }
+    if (instruction->argobjex_cache != mp_const_none) {
+        mp_printf(&mp_plat_print, " #");
+        mp_obj_print_helper(&mp_plat_print, instruction->argobjex_cache, PRINT_REPR);
+    }
+
+    mp_printf(&mp_plat_print, "\n");
+}
+
+#endif // PROF_PRINT_INSTR
+
+uint prof_bytecode_lineno(const mp_raw_code_t *rc, size_t bc) {
+    const mp_bytecode_prelude_t *prelude = &rc->prelude;
+    const byte *ip = prelude->line_info;
+    size_t source_line = 1;
+    size_t c;
+
+    bc = (bc + (prelude->bytecode - prelude->code_info)) - prelude->code_info_size;
+
+    while ((c = *ip)) {
+        size_t b, l;
+        if ((c & 0x80) == 0) {
+            // 0b0LLBBBBB encoding
+            b = c & 0x1f;
+            l = c >> 5;
+            ip += 1;
+        } else {
+            // 0b1LLLBBBB 0bLLLLLLLL encoding (l's LSB in second byte)
+            b = c & 0xf;
+            l = ((c << 4) & 0x700) | ip[1];
+            ip += 2;
+        }
+        if (bc >= b) {
+            bc -= b;
+            source_line += l;
+        } else {
+            // mp_printf(&mp_plat_print, "found line %d %d %d\n", source_line, bc, b);
+            // found source line corresponding to bytecode offset
+            break;
+        }
+    }
+
+    return source_line;
+}
+
+uint prof_bytecode_linenox(const byte *line_info, size_t bc) {
+    const byte *ip = line_info;
+    size_t source_line = 1;
+    size_t c;
+
+    while ((c = *ip)) {
+        size_t b, l;
+        if ((c & 0x80) == 0) {
+            // 0b0LLBBBBB encoding
+            b = c & 0x1f;
+            l = c >> 5;
+            ip += 1;
+        } else {
+            // 0b1LLLBBBB 0bLLLLLLLL encoding (l's LSB in second byte)
+            b = c & 0xf;
+            l = ((c << 4) & 0x700) | ip[1];
+            ip += 2;
+        }
+        if (bc >= b) {
+            bc -= b;
+            source_line += l;
+        } else {
+            // mp_printf(&mp_plat_print, "found line %d %d %d\n", source_line, bc, b);
+            // found source line corresponding to bytecode offset
+            break;
+        }
+    }
+
+    return source_line;
+}
+
+void prof_extract_prelude(const byte *bytecode, mp_bytecode_prelude_t *prelude) {
```

Comment:
Is it possible to reuse `extract_prelude` from `persistentcode.c`?

Reply:
It is not possible to reuse the `extract_prelude` as it is right now. It would have to be changed and I would like to keep this PR as decoupled as possible while in the WIP mode. I would make such a change in a later stage if it will make sense.

Reply:
Ok, makes sense.